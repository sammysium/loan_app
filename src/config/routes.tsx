import { createNativeStackNavigator, NativeStackNavigationProp } from '@react-navigation/native-stack';
import ApplyForLoanScreen from '@views/screens/ApplyForLoanScreen';
import LoansScreen from '@views/screens/LoansScreen';

type RootStackParamList = {
    Loans: undefined;
    Apply: undefined
};

export type NavigationProps = NativeStackNavigationProp<RootStackParamList, 'Loans'>;


const Stack = createNativeStackNavigator<RootStackParamList>();

const routes = () => {
    return <>
        <Stack.Navigator initialRouteName="Loans">
            <Stack.Screen name="Loans" component={LoansScreen} options={{
                headerShown: false
            }} />

            <Stack.Screen name="Apply" component={ApplyForLoanScreen} options={{
                headerShown: false
            }} />
           
        
        </Stack.Navigator>
    </>
}

export default routes;


