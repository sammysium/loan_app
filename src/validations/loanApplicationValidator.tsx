import GenericFormErrorFields from "@models/GenericFormErrorFields"
import { hasNoValue, isValidEmail, isValidMoney } from "./formValidator"
import ILoanApplication from "@models/ILoanApplication"
import { translator } from "@utils/translator"
import ILoan from "@models/ILoan"
import { config } from "@config/config"

export const validateLoanApplicationData = (data: ILoanApplication): GenericFormErrorFields => {
    const errorMessages: GenericFormErrorFields = {}
    console.log('%%%%%%% ', data)
    if (!isValidEmail(data.email)) {
        errorMessages["email"] = translator({translationId: "errorInvalidEmailFormat"})
    }

    if (hasNoValue(data.full_name)) {
        errorMessages["full_name"] = translator({translationId: "errorFullNameMissing"})
    }

    if (!isValidMoney(data.loan_amount, config.default.loan.minimum , config.default.loan.maximum)) {
        errorMessages["loan_amount"] = translator({translationId: "errorInvalidLoanAmount"})
    }

    if (hasNoValue(data.loan_purpose)) {
        errorMessages["loan_purpose"] = translator({translationId: "errorInvalidPurpose"})
    }

    return errorMessages;
}