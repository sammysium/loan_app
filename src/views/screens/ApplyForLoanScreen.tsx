import ILoan from "@models/ILoan";
import MainLayout from "@views/layouts/MainLayout";
import { Alert, ScrollView, StyleSheet, View } from "react-native";
import { Text, useTheme } from "react-native-paper";
import { translator } from "@utils/translator";
import { useState } from "react";
import { NavigationProps } from "@config/routes";
import { globalStyles } from "@views/themes/styles";
import ILoanApplication from "@models/ILoanApplication";
import { validateLoanApplicationData } from "@validations/loanApplicationValidator";
import ServiceLoans from "@services/ServiceLoans";
import GenericFormErrorFields from "@models/GenericFormErrorFields";
import TextEntry from "@views/components/TextEntry";
import AppButton, { ButtonType } from "@views/components/AppButton";
import Spacer from "@views/components/Spacer";
import { useStore } from "@store/useStore";

interface IProps {
    navigation: NavigationProps
}

const ApplyForLoanScreen: React.FC<IProps> = ({ navigation }) => {
    
    const { colors } = useTheme()
    const [errors, setFormErrors] = useState<GenericFormErrorFields>({})
    const [loading, setLoading] = useState(false)
    const [loanApplicationData, setLoanApplicationData] = useState<ILoanApplication>(
        {
            full_name: "",
            email: "",
            loan_amount: 0,
            loan_purpose: ""
        }
    )

    const handleEntries = (fieldName: string, value: string | number) => {
        setLoanApplicationData({
            ...loanApplicationData, [fieldName]: value
        })
    }


    const submitData = async () => {
        setLoading(true)
        //validate it all first
        const validationErrors = validateLoanApplicationData(loanApplicationData, {id: 1, name: "", maximumAmount: 100, interestRate: 3})
        console.log('---- error --- ', validationErrors)
        setFormErrors(validationErrors)
        if (Object.keys(validationErrors).length === 0) {
            // send the loan application
            try {
                const result = await ServiceLoans.submitNewLoan(loanApplicationData)
                console.log(58, result)
                Alert.alert(
                    translator({translationId:"success"}), 
                    translator({translationId:"applicationSubmitted"})
                )
                navigation.navigate("Loans")
            } catch (error) {
                
                Alert.alert(
                    translator({translationId:"error"}), 
                    translator({translationId:"applicationSubmissionFailed"})
                )

            } finally {
                
                    setLoading(false)
                
              
            }
        } else {
            setLoading(false)
        }

    }

    return <MainLayout navigation={navigation}>
        <ScrollView style={styles.container}>
            <Text style={globalStyles.medium_header}>{translator({ translationId: "applyForLoan" })}</Text>

            <TextEntry onChange={(e) => handleEntries("full_name", e)}
                label={translator({ translationId: "fullName" })}
                placeHolderText={translator({ translationId: "fullName" })}
                errorMessage={errors["full_name"]}
                testId="fullName"
            />

            <TextEntry 
                onChange={(e) => handleEntries("email", e)}
                label={translator({ translationId: "email" })}
                placeHolderText="yourname@example.com"
                errorMessage={errors["email"]}
                testId="email"
                keyboardType="email-address"
            />

            <TextEntry 
                onChange={(e) => handleEntries("loan_amount", parseFloat(e))}
                label={translator({ translationId: "loanAmount" })}
                placeHolderText="UGX"
                errorMessage={errors["loan_amount"]}
                testId="loanAmount"
                keyboardType="numeric"
            />

            <TextEntry
                onChange={(e) => handleEntries("loan_purpose", e)}
                label={translator({ translationId: "loanPurpose" })}
                placeHolderText={translator({ translationId: "loanPurposeExample" })}
                errorMessage={errors["loan_purpose"]}
                testId="loanPurpose"
            />

            <AppButton
                backgroundColor={colors.primary}
                upperCase={true}
                buttonType={ButtonType.Contained}
                labelTranslationId={!loading ? 'applyForLoan': 'submittingApplication'}
                textColor={colors.secondaryContainer}
                onPressHandler={submitData}
                disabled={loading}
                loading={loading}
                testId="submitLoanApplicationButton"
            />
            <Spacer vertical={15}/>
        </ScrollView>
    </MainLayout>
}

const styles = StyleSheet.create({
    container: {
        flex:1,
       
        paddingBottom: 40,
    },
    row: {
        justifyContent: 'space-between',
    },
    contentContainer: {
        flexGrow: 1,
    },
});

export default ApplyForLoanScreen;