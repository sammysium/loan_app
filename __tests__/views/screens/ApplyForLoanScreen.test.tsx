import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react-native';
import { Alert } from 'react-native';
import { NavigationProps } from '@config/routes';
import * as loanApplicationValidator from '@validations/loanApplicationValidator';
import ApplyForLoanScreen from '@views/screens/ApplyForLoanScreen';


jest.mock('@utils/translator')

// Mock Alert
jest.spyOn(Alert, 'alert').mockImplementation(() => { });

// Create a mock navigation object that includes required methods
const mockNavigate = jest.fn();
const mockGoBack = jest.fn();
const mockReset = jest.fn();
const mockIsFocused = jest.fn();
const mockDispatch = jest.fn();

const mockNavigation: NavigationProps = {
    navigate: mockNavigate,
    goBack: mockGoBack,
    reset: mockReset,
    isFocused: mockIsFocused,
    dispatch: mockDispatch,
    setParams: jest.fn(),
    addListener: jest.fn(),
    removeListener: jest.fn(),
    canGoBack: jest.fn(),
    getState: jest.fn(),
    getParent: jest.fn(),
    setOptions: jest.fn(),
    getId: jest.fn(),
    replace: jest.fn(), push: jest.fn(), pop: jest.fn(), popToTop: jest.fn()
};

describe('ApplyForLoanScreen', () => {

    let spyAlert = jest.spyOn(Alert, 'alert');

    jest.spyOn(loanApplicationValidator, 'validateLoanApplicationData').mockImplementation(() => ({}));

    beforeEach(() => {

        jest.restoreAllMocks();
        spyAlert = jest.spyOn(Alert, 'alert');
    });

    afterEach(() => {

        spyAlert.mockReset();

    });

    it('should match snapshot', () => {
        const { toJSON } = render(<ApplyForLoanScreen navigation={mockNavigation} />);
        expect(toJSON()).toMatchSnapshot();
    });


    it('should process form', async () => {

        const { getByTestId } = render(<ApplyForLoanScreen navigation={mockNavigation} />);


        const fullNameInput = getByTestId("fullName");

        const emailInput = getByTestId('email');
        const loanAmountInput = getByTestId('loanAmount');
        const loanPurposeInput = getByTestId("loanPurpose");
        const submitButton = getByTestId("submitLoanApplicationButton")

        fireEvent.changeText(fullNameInput, 'John Doe');
        fireEvent.changeText(emailInput, 'john@example.com');
        fireEvent.changeText(loanAmountInput, '1000');
        fireEvent.changeText(loanPurposeInput, 'Business Expansion');

   

         act(() => {
        
            fireEvent.press(submitButton)        
        });

    });

});
