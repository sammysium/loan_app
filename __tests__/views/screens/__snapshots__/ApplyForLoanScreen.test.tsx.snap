// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplyForLoanScreen should match snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#F3FBFB",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderBottomColor": "#ddd",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "height": 60,
        "justifyContent": "space-between",
        "paddingHorizontal": 15,
        "width": 750,
      }
    }
  >
    <Text
      style={
        {
          "color": "#B7E3F1",
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Numida
    </Text>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(232, 222, 248, 1)",
            "borderRadius": 6,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="appButton-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(232, 222, 248, 1)",
              "borderColor": "transparent",
              "borderRadius": 6,
              "borderStyle": "solid",
              "borderWidth": 0,
              "flex": undefined,
              "flexDirection": "row",
              "justifyContent": "center",
              "minWidth": 64,
              "paddingHorizontal": 1,
              "paddingVertical": 1,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="appButton-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 6,
                },
              ]
            }
            testID="appButton"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 12,
                        },
                        false,
                        false,
                        {
                          "color": "rgba(103, 80, 164, 1)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        {},
                      ],
                    ],
                  ]
                }
                testID="appButton-text"
              />
            </View>
          </View>
        </View>
      </View>
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(103, 80, 164, 1)",
            "borderRadius": 6,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="appButton-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(103, 80, 164, 1)",
              "borderColor": "transparent",
              "borderRadius": 6,
              "borderStyle": "solid",
              "borderWidth": 0,
              "flex": undefined,
              "flexDirection": "row",
              "justifyContent": "center",
              "minWidth": "auto",
              "paddingHorizontal": 1,
              "paddingVertical": 1,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="appButton-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 6,
                },
              ]
            }
            testID="appButton"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        {
                          "marginHorizontal": 8,
                        },
                        {
                          "textTransform": "uppercase",
                        },
                        {
                          "color": "rgba(232, 222, 248, 1)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        {
                          "fontSize": 9,
                          "marginHorizontal": 20,
                          "marginVertical": 2,
                        },
                      ],
                    ],
                  ]
                }
                testID="appButton-text"
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flex": 1,
        "paddingHorizontal": 10,
      }
    }
  >
    <RCTScrollView
      style={
        {
          "flex": 1,
          "paddingBottom": 40,
        }
      }
    >
      <View>
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "color": "#000",
                "fontSize": 20,
                "fontWeight": "600",
                "marginBottom": 7,
              },
            ]
          }
        />
        <View
          style={
            {
              "flexDirection": "column",
              "marginBottom": 10,
            }
          }
        >
          <TextInput
            keyboardType="default"
            onChangeText={[Function]}
            style={
              {
                "borderColor": "blue",
                "borderRadius": 4,
                "borderWidth": 1,
                "padding": 8,
              }
            }
            testID="fullName"
          />
        </View>
        <View
          style={
            {
              "flexDirection": "column",
              "marginBottom": 10,
            }
          }
        >
          <TextInput
            keyboardType="default"
            onChangeText={[Function]}
            placeholder="yourname@example.com"
            style={
              {
                "borderColor": "blue",
                "borderRadius": 4,
                "borderWidth": 1,
                "padding": 8,
              }
            }
            testID="email"
          />
        </View>
        <View
          style={
            {
              "flexDirection": "column",
              "marginBottom": 10,
            }
          }
        >
          <TextInput
            keyboardType="numeric"
            onChangeText={[Function]}
            placeholder="UGX"
            style={
              {
                "borderColor": "blue",
                "borderRadius": 4,
                "borderWidth": 1,
                "padding": 8,
              }
            }
            testID="loanAmount"
          />
        </View>
        <View
          style={
            {
              "flexDirection": "column",
              "marginBottom": 10,
            }
          }
        >
          <TextInput
            keyboardType="default"
            onChangeText={[Function]}
            style={
              {
                "borderColor": "blue",
                "borderRadius": 4,
                "borderWidth": 1,
                "padding": 8,
              }
            }
            testID="loanPurpose"
          />
        </View>
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgba(103, 80, 164, 1)",
              "borderRadius": 6,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="submitLoanApplicationButton-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "rgba(103, 80, 164, 1)",
                "borderColor": "transparent",
                "borderRadius": 6,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "flexDirection": "row",
                "justifyContent": "center",
                "minWidth": 64,
                "paddingHorizontal": 1,
                "paddingVertical": 1,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="submitLoanApplicationButton-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 6,
                  },
                ]
              }
              testID="submitLoanApplicationButton"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        [
                          {
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          false,
                          {
                            "marginHorizontal": 24,
                            "marginVertical": 10,
                          },
                          false,
                          {
                            "textTransform": "uppercase",
                          },
                          {
                            "color": "rgba(232, 222, 248, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          {},
                        ],
                      ],
                    ]
                  }
                  testID="submitLoanApplicationButton-text"
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            [
              {
                "height": 15,
                "width": 0,
              },
            ]
          }
        />
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
